src/System.Collections/src/System/Collections/BitArray.cs

```diff
@@ -366,23 +366,32 @@ public int Length
 
                 if (value > m_length)
                 {
-                    // clear high bit values in the last int
-                    int last = GetArrayLength(m_length, BitsPerInt32) - 1;
-                    int bits = m_length % 32;
-                    if (bits > 0)
-                    {
-                        m_array[last] &= (1 << bits) - 1;
-                    }
-
-                    // clear remaining int values
-                    Array.Clear(m_array, last + 1, newints - last - 1);
+                    TrimUpperBits();
                 }
 
                 m_length = value;
-                _version++;
             }
         }
 
+        private void TrimUpperBits()
+        {
+            // clear high bit values in the last int
+            int last = GetArrayLength(m_length, BitsPerInt32) - 1;
+            int bits = m_length % 32;
+            if (bits > 0)
+            {
+                m_array[last] &= (1 << bits) - 1;
+            }
+
+            // clear remaining int values
+            if (last + 1 < m_array.Length)
+            {
+                Array.Clear(m_array, last + 1, m_array.Length - last - 1);
+            }
+
+            _version++;
```

Comment:
It seems to me that the culprits for the excessive bits are three: `BitArray(int, bool)` in case the second argument is `true`, `SetAll(bool)` when the argument is `true`, and `Not()`. Since the list is short, it might make sense to just add the trimming there. This solution has the benefit of always keeping even the internal data structure "correct" at all times.

The downside to this approach is that if the API surface is extended (e.g. with more bitwise operations), the implementer must take care to add the trimming wherever necessary.